### PR TITLE
Operator e2e Fix

### PR DESCRIPTION
**Pull request by irisdingbj**
bases on https://github.com/istio/operator/pull/404

@sdake I am trying to fix the e2e failure for operator controller here based on your above PR.

### CODE REVIEW

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: this part is suspect for prow.  I'd suggest taking a look at removal of this for the prow case (but keep for the e2e case).  Additionally I have read metallb does nt work on Mac,but does work on Linux, so we may need some host checking before using metallb and then use `-use-local` or whatever the test flag is for the actual e2e run based upon that conditional.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: this part looks correct.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: surprised this merged with sdake as my repo.  I was using that for testing only and have since overwritten my master.  github.com/istio/istio.git contains all of the changes needed for this PR to work - so this line looks correct.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: please delete this echo line - it is no longer needed as I added YAML terminators in a prior PR to all of the yaml files.  I am not sure if my master PR was backported - if not it needs to be.  I will look at this now and make it so.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: `--use-local-cluster=true` is fine for some conditions (not using metallb) - but I really do want to test load-balancer cases in the 2e test cases locally.  I am good with this minor change (adding metallb into the system) after we get a good baseline).

Comment (sdake) [reply tests/e2e/e2e.sh]: iirc make e2e_simple_run does not build istioctl - which is mandatory for the test case to run - but I could be mistaken (this is why I build it in the original work).

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: not sure what this is or does.  Did you add prow/lib.sh to the repo?  If its already in repo - make sure it does what you expect it to. @howardjohn and I were speaking of the need to generalize this work so it can be reused around all satellite repos (cni and operator to be specific at first) - although that generalization work can follow on making it work for the specific case of operator - which is essential for the release.  If your struggling, I should be available for a conversation tonight on webex.

**Review by sdake — COMMENTED** [review]
looks good - minus the loss of metallb functionality.  If removal of metallb gets things moving for e2e, can you follow up this work with a metallb implementation for folks running on Linux systems?  After that, a further generalization of this script so that it works correctly for cni is in order - perhaps you can take that work on as a followup.

Cheers
-steve

Cheers
-steve

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: Please be aware - `make e2e` should work locally on your system (if your on Linux)l.  I didn't test mac local e2e, but that must work, although Linux dev environments take priority.  `make e2e` was working locally and in prow for me prior to this submission of this PR, however, I noticed I was using master of sdake/istio - and may have had some minor changes in istio/istio that were not pushed (and are now sady lost as I have force pushed to sdake/istio)...

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: @istio/wg-environments-maintainers please take care when reviewing PRs to not **approve** PRs that contain personal git repositories.  Thewse personal git repositories may be used for testing if they are from trusted developers, however, they should not be merged and testing should not be permitted for folks new to the project (secrets can be exposed...)

Cheers
-steve

**Review by sdake — COMMENTED** [review]
looks pretty good, some more comments inline.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: It is part of istio repo. So we can resue all the functions there.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: THis can work on my mac locally

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: setup_kind_cluster will take care all of this including cleanup existed kind cluster.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: Fixed!

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: this should just be `make`.  Make determines if it is running in a container environment or not, and this model would break container builds.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: can you explain this in more detail.  This doesn't make immediate sense.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: the ordering shouldn't matter - not sure this change is correct.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: disagree - operator should stand alone in its dependencies on building the repository.  If we need a common files prow.ib, we can do that, but it should be merged as a separate PR in common-files repo (see link above).

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: `--test.run . TestSimple1` is a subset of the test cases that should be run.  Recommend removing and fixing the deficiencies in istio/istio or this repo.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: It will fails if we deploy the icp cr a bit ahead. It complains that no crd found. By making cr deploy the last one, we can make sure no such problem happens.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: Let's first make sure our e2e can run.  This will not make the operator build dependent on istio. This is just a test dependency as we will depend on istio for e2e test anyway.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: t will fails if we deploy the icp cr a bit ahead. It complains that no crd found. By making cr deploy the last one, we can make sure no such problem happens

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: rename it as `ROOT ` to prevent confusion

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: Changed to `make controller docker` as we only need these two targets.

Comment (sdake) [reply tests/e2e/e2e.sh]: `make docker controller` are or should be dependencies of the top level `Makefile.core.mk` file so they build whenever e2e runs.  You may need to specify docker and docker.save as `PHONY` target if not already done.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: it is better to accept these from the `Makeile.core.mk` file directly rather than have to maintain them in two places.  On the call to `e2e.sh` i'd recommend HUB=$(HUB) TAG=$(TAG) tests/e2e/e2e.sh such that user overrides of HUB or TAG are used during the build process.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: try with `HUB=$(HUB)

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: replace istio-testing/oeprator:1.5-dev with `$(HUB)/operator:$(TAG)

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: rename function to `build_istio_images()`

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: I believe this may not be necessary as it was not necessary prior for e2e testing to work.  Perhaps this was a side-effect of running the `make e2e` test target in the istio repository - but there should be no strong rationale to duplicate this functionality between repositories (it will make maintenance harder).

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: I continue not to like this use of (cd ../../../; pwd).  There is an environment variable (IIRC) that contains `PWD` in the top level Makefile.  Can you see if that can be used?  If its not being passed to `Makefile.core.mk` - it can be easily added.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: I don't see a setup_kind_cluster in this file.  It seems to have been removed.  Is this being imported from the istio repository in some way?  If so, please duplicate the work here.  While duplication is generally bad, we have a process for managing duplication consistently across repositories using https://github.com/istio/common-files.  I would far prefer this common-files approach to distribute a prow library then calling one prow library on one place and having delta code in a different place (reference cni e2e where this was copied from).

I'd like to get strong baselines (in this PR) for all of the various repositories before working on the unification of the kind handling (in a later PR).

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: imo setup shouldn't delete, it should setup.  If we need to call delete to get the system progressing, thats fine.  i recommend not confusing the function names with what the functions do.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: getting metallb to work on Linux was hard, but I can revive this work later if need be.  Please open a separate PR and place a hold on it to add metlalb back to the system and I can take on reimplementation from there within the PR queue.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: if metallb works locally on your mac without ``--use-local-cluster=true` that would be ideal.  Can you clarify your statement?

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: this is definately wrong - although it may be my original work where it was wrong.

I'd recommend trying `ISTIO_DIR=$(mktemp -d)`.  If that fails because of problems in Istio, we will need to work through that problem (we can do that after this PR merges) because hardcode paths in the build system are undesireable.

Comment (sdake) [reply tests/e2e/e2e.sh]: Further following this up - the git clone below could negatively interact with a local copy of istio (I recognize a test for the directory is done) but this would result in $ISTIO/istio being used from the local filesystem if the developer has work in that location.  This is highly undesirable for local test modes.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: thats bad news - but not unexpected.  Thanks.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: see review for more information about why using a prow library is an undesirable approach.  As painful as it may be, please copy the parts you need into this e2e.sh script.  As followup work a refactor of the e2e for the various repos (istio/operator, istio/cni, istio/istio) can be done in https://github.com/istio/common-files.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: this should use the same environment variables as the rest of the build system `HUB=$(HUB) TAG=$(TAG) make docker.app`

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: see comment on line 64 about hub and tag usage.

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: why pop to push again?  Is the need to call setup_docker?

**Review by sdake — COMMENTED** [review]
implementation looks ok.  There are some problems around HUB,TAG variable usage and the git cloing specifically.  If you tell me it wont work with a git clone in a temp dir, we can merge without that.

I'd really strongly like to avoid the usage of the prow library - but instead copy the relevant bits into this script so we can make a unified e2e script in common-tools as a followup PR.

**Comment by sdake** [comment]
@howardjohn would you mind reviewing?

File: tests/e2e/e2e.sh
Comment (howardjohn) [reply tests/e2e/e2e.sh]: It needs to be `src/istio.io/istio`, since prow ~~will~~ should check out istio/istio for us instead of doing the clone here. 
And to your other concern, I think the clone only happens if the folder doesn't exist

File: tests/e2e/e2e.sh
Comment (howardjohn) [inline tests/e2e/e2e.sh]: I don't think we should explicitly set the node here, or this will break when we update kind. Really, any of this can break if we update ./prow/lib.sh so its already sort of at risk

**Review by howardjohn — COMMENTED** [review]
This is pretty misleading, you are not actually testing the PR

You set the HUB and TAG, build docker images, but they are **not** used.

When you pass HUB and TAG to the test, it is not used because the operator part of the test doesn't support it. 

When you render the operator install yaml, it has the hardcoded hub and tag: https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_operator/456/e2e_operator/115/artifacts/simple-auth-test-a9a5b04cf9db4c37a5fe/yaml/istio-operator.yaml

So anything you are doing with docker and hubs and tags is entirely unused as far as I can tell

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: ok thats fair.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: @howardjohn right, but if the repo does exist, then the e2e is done based upon the local version of Istio?  I agree with your other points.  Not sure what to do with the prefix `$(ROOT}`

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: right, my suggestion was to keep this per repo - and then follow this work up with a common-files replication.  agree re node, we should bake node into the buildtools image imo.

**Comment by howardjohn** [comment]
+1 We should bake the default node image into the build. I tried
unsuccessfully

**Comment by irisdingbj** [comment]
> This is pretty misleading, you are not actually testing the PR
> 
> You set the HUB and TAG, build docker images, but they are **not** used.
> 
> When you pass HUB and TAG to the test, it is not used because the operator part of the test doesn't support it.
> 
> When you render the operator install yaml, it has the hardcoded hub and tag: https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_operator/456/e2e_operator/115/artifacts/simple-auth-test-a9a5b04cf9db4c37a5fe/yaml/istio-operator.yaml
> 
> So anything you are doing with docker and hubs and tags is entirely unused as far as I can tell

good catch! Thanks. I will fix this.

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: removed already

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: removed the node images here

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: revised

File: tests/e2e/e2e.sh
Comment (howardjohn) [inline tests/e2e/e2e.sh]: where is app coming from?

File: tests/e2e/e2e.sh
Comment (howardjohn) [inline tests/e2e/e2e.sh]: the hub should not be a real hub, or jts impossible to tell whether we are using the local image or accidentally using a published image

File: tests/e2e/e2e.sh
Comment (howardjohn) [inline tests/e2e/e2e.sh]: nit - we should probably using gcr for the hub for all Istio components, and then the local image for the operator. We aren't testing istio/istio here so no need to build those images

File: tests/e2e/e2e.sh
Comment (howardjohn) [reply tests/e2e/e2e.sh]: nevermind, I see we build it later, but probably don't need to

**Review by howardjohn — COMMENTED** [review]
looks good, just a couple minor comments. I think this is in a good state though

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: we need this for simple testing, especially the testNgix part

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: just need to build istio-testing/app:istio-testing as we will not build this images during normal istio build

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: Need to set this as a real hub otherwise we need to modify deploy/operator.yaml..

**Comment by irisdingbj** [comment]
> looks good, just a couple minor comments. I think this is in a good state though

Thanks @howardjohn seems it fails for `istioctl kube-inject this time..` not sure the reason yet

File: tests/e2e/e2e.sh
Comment (howardjohn) [reply tests/e2e/e2e.sh]: I think We should not be building it, we should use gcr.io/istio-testing/app:latest. We aren't building pilot or galley, so why build the test app?

File: tests/e2e/e2e.sh
Comment (howardjohn) [reply tests/e2e/e2e.sh]: I think it's better to modify that. Being able to change the hub/tag is a requirement for users, so we need to support it some way or another.

Otherwise, it's likely someone changes the image pull policy of something else that causes us to not use the local image and we don't notice?

**Comment by howardjohn** [comment]
the istioctl inject is not related to this pr, I have a PR on istio/istio to fix it

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: agree!  Let me put the operator.yaml one copy into testdata

File: tests/e2e/e2e.sh
Comment (irisdingbj) [reply tests/e2e/e2e.sh]: > the istioctl inject is not related to this pr, I have a PR on istio/istio to fix it

Thanks for the info!

Comment (irisdingbj) [reply tests/e2e/e2e.sh]: changed to use gcr.io/istio-testing/app:latest and skip building istio-testing/app

**Comment by irisdingbj** [comment]
https://github.com/istio/istio/pull/18352

File: deploy/operator.yaml
Comment (sdake) [inline deploy/operator.yaml]: @ostromart I believe I like Always as a pull policy better.  I understand why this PR changes the default pull policy.  If Always is specified, docker will pull the image in all cases, even though it may not be rermotely pushed.

A sed operation on this line may make sense.  Do you have any thoughts on what is bettter?

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: will address in a followon PR if we can get this PR merged today.

File: tests/e2e/e2e.sh
Comment (sdake) [reply tests/e2e/e2e.sh]: @howardjohn can you clarify your suggestion here?

File: tests/e2e/e2e.sh
Comment (sdake) [inline tests/e2e/e2e.sh]: this HUB in use doesn't match the HUB at line 26.

File: tests/e2e/testdata/operator.yaml
Comment (sdake) [inline tests/e2e/testdata/operator.yaml]: why not use sed to generate this file on the fly as to avoid duplication of the manifests which may become inconsistent over time?

**Review by sdake — APPROVED** [review]
This is pretty close and I feel comfortable merging assuming the unresolved issues are addressed in a followup iterative PR.  Assuming this PR can pass the checks, we can merge it now and follow up with fixes for the outstanding issues.  If this PR still blocks in the checks, @irisdingbj can you get it passing as soon as viable?

**Comment by sdake** [comment]
groan - I had forgot the e2e tests were non-voting.  I'll take a look at fixing this up today - if not @irisdingbj - can you take a look at fixing when you start your day.

Cheers
-steve

**Pull request merged**
